applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp

```diff
@@ -0,0 +1,163 @@
+    if (TensileStrength > result[0]) return result;
+
+    result[0] = (TensileStrength - Cohesion * std::cos(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    result[1] = (Cohesion * std::cos(FrictionAngle) - TensileStrength * std::sin(FrictionAngle)) /
+                (1.0 - std::sin(FrictionAngle));
+    return result;
+}
+
+bool IsStressAtTensionApexReturnZone(const Vector& rTrialSigmaTau, double TensileStrength, double Apex)
+{
+    return TensileStrength < Apex && rTrialSigmaTau[0] - rTrialSigmaTau[1] - TensileStrength > 0.0;
+}
+
+bool IsStressAtTensionCutoffReturnZone(const Vector& rTrialSigmaTau, double TensileStrength, double Apex, const Vector& rCornerPoint)
+{
+    return TensileStrength < Apex &&
+           rCornerPoint[1] - rTrialSigmaTau[1] - rCornerPoint[0] + rTrialSigmaTau[0] > 0.0;
+}
+
+bool IsStressAtCornerReturnZone(const Vector& rTrialSigmaTau, double DilatancyAngle, const Vector& rCornerPoint)
+{
+    return rTrialSigmaTau[0] - rCornerPoint[0] - (rTrialSigmaTau[1] - rCornerPoint[1]) * std::sin(DilatancyAngle) >= 0.0;
+}
+
+Vector ReturnStressAtTensionApexReturnZone(double TensileStrength)
+{
+    Vector result{2};
+    result[0] = TensileStrength;
+    result[1] = 0.0;
+    return result;
+}
+
+Vector ReturnStressAtTensionCutoffReturnZone(const Vector& rSigmaTau,
+                                             const Vector& rDerivativeOfFlowFunction,
+                                             double        TensileStrength)
+{
+    const auto lambda = (TensileStrength - rSigmaTau[0] - rSigmaTau[1]) /
+                        (rDerivativeOfFlowFunction[0] + rDerivativeOfFlowFunction[1]);
+    return rSigmaTau + lambda * rDerivativeOfFlowFunction;
+}
+
+Vector ReturnStressAtCornerReturnZone(const Vector& rCornerPoint)
+{
+    Vector result{2};
+    result[0] = rCornerPoint[0];
+    result[1] = rCornerPoint[1];
+    return result;
+}
+
+Vector ReturnStressAtRegularFailureZone(const Vector& rSigmaTau,
+                                        const Vector& rDerivativeOfFlowFunction,
+                                        double        FrictionAngleInRadians,
+                                        double        Cohesion)
+{
+    const auto cof1      = std::sin(FrictionAngleInRadians);
+    const auto cof2      = Cohesion * std::cos(FrictionAngleInRadians);
+    const auto numerator = cof1 * rDerivativeOfFlowFunction[0] + rDerivativeOfFlowFunction[1];
+    const auto lambda    = (cof2 - rSigmaTau[0] * cof1 - rSigmaTau[1]) / numerator;
+    return rSigmaTau + lambda * rDerivativeOfFlowFunction;
+}
+
+} // namespace
+
+namespace Kratos
+{
+CoulombWithTensionCutOffImpl::CoulombWithTensionCutOffImpl(double FrictionAngleInRad,
+                                                           double Cohesion,
+                                                           double DilatationAngleInRad,
+                                                           double TensileStrength)
```

Comment:
Some renaming suggestions:
```suggestion
CoulombWithTensionCutOffImpl::CoulombWithTensionCutOffImpl(double FrictionAngleInRadians,
                                                           double Cohesion,
                                                           double DilatancyAngleInRadians,
                                                           double TensileStrength)
```